### ISSUE TITLE
Object literal with a key of a subtype of string not inferable as Record

### ISSUE THREAD

**Issue by aloifolia** [microsoft/TypeScript#53912]
I have the following function:

```ts
function createRecord1<X extends string>(x: X): Record<X, number> {
  return {[x]: 42};
}
```

The compiler says 

```ts
Type '{ [x: string]: number; }' is not assignable to type 'Record<X, number>'.(2322)
```

However, the following functions do not show any error.

```ts
function createRecord2<X extends number>(x: X): Record<X, number> {
  return {[x]: 42};
}

function createRecord3<X extends symbol>(x: X): Record<X, number> {
  return {[x]: 42};
}
```

Why is there a difference? I know that in the first example I can get rid of the error with a simple as-Typecast, but I'd like not to sacrifice the type safety.

Here is a [playground with all functions](https://www.typescriptlang.org/play?ts=5.0.4#code/GYVwdgxgLglg9mABBATgUwIZTQJTROFAEwEYAeADUTQA9swiBnRRqFGMAcwD4AKGgFyIKASiF4CxSgBpEYEAFsARmhTdEAbwBQiROiggUSDQG0aAXSEAWAEwBfANxa7WraEiwEydFlz5CRDaU1HRoDMzyyqp8gsJiiBIBMnKKKmqaOnpoBkaaZpaIto7Oru7Q8EiomNiJxADMwbT0TCwAnspwADYxQqLi-lIUspFp6tq6+obG+db2Ti5AA).

**Comment by RyanCavanaugh** [comment]
This is a correct error, since a legal call is
```ts
const x = createRecord<"A" | "B">("A")
```
but `x` does not have both `A` and `B` properties

**Comment by fatcerberus** [comment]
> What am I missing? 🤔

`Record<"A” | "B", number>` must have both properties, but `createRecord<"A" | "B">("A”)` will return an object with only one of them at runtime.

**Comment by aloifolia** [comment]
OK, that sounds reasonable. But why does the compiler give different results when you have numbers or symbols as keys instead of strings? It's specifically the difference between the 3 examplary functions that puzzle me.

**Comment by potzkovge** [comment]
> > What am I missing? 🤔
> 
> `Record<"A” | "B", number>` must have both properties, but `createRecord<"A" | "B">("A”)` will return an object with only one of them at runtime.

It says the constraint for the key of a Record is
![image](https://user-images.githubusercontent.com/3837678/233160573-d3486a01-dd82-430e-b7f7-86f8623c3903.png)
His example shows it behaving as expected for number and symbol but not for string.
What you describe is a specific type which obviously works but the value is static the question is why the generic is not working.

[Playground Example](https://www.typescriptlang.org/play?ts=5.0.4#code/GYVwdgxgLglg9mABBATgUwIZTQJTROFAEwFEUVCAeADQD4AKADwC5FqBKVvA4mgGkRgQAWwBGaFLUQBvAFCJE6KCBRJpAbUYBdVgBYATAF8A3LMOzZUAJ4AHNIgDCWAHIZh9gLyIARMJjBgK28LUEhYBGR0LFx8QiIAZSgsGAgmVicoV3dORG44ygystAEhMQkpOQUlFTV5BURNHUQDOvNzWQIwAGcoRVjiROSIRC9UTGw8gaTYVN9-QO92DoQuuAAbNAA6NbgAc3oAA0GZxB3d1gASaXQeBOmU9TmAoK1DA-ZEAHpP5v0Lb8QAAV0AA3eAgLqINCMNw2DaQ4CERC7NBgCQpQQicQoT5dKxidYhcDQeBIMbRSZEfQ0KGMbBgIiQ0rYhgsNg5Sn8TFlSQyOrVVQyRp6IymdqdHp9W76EaRcYxaX0fRLCXrLZnQ7OLESU57S7XfpU9T6V7vL4-FqyUIkiLkiaGgDMNOh9MZiDxBLWrNYHC4hq5zPKfKqaGUgo02hFJjMFglvS6AEZZfF8aJ1vQVStejc4g7ZXaFbn6InM901ds9ocU57dedEFcc8QHepE6aPgCDEA)

**Comment by fatcerberus** [comment]
It doesn’t work because you’re not returning a value which is valid for all possible types of `X`.  You seem to be very confused and this is not a support forum; I suggest taking this to Discord or Stack Overflow.  There is no TypeScript bug here.

**Comment by potzkovge** [comment]
> It doesn’t work because you’re not returning a value which is valid for all possible types of `X`. You seem to be very confused and this is not a support forum; I suggest taking this to Discord or Stack Overflow. There is no TypeScript bug here.

True but neither is number nor symbol. If this not regarded as a bug then another platform for further clarification might be needed. Thanks.

**Comment by aloifolia** [comment]
@potzkovge The example given by @RyanCavanaugh makes sense in my eyes, as well. However, please have a look at my original question: why is there a difference between strings, numbers and symbols?

If the string case (createRecord1) is correct, the others (createRecord2 and createRecord3) must apparently be incorrect. Or not?

**Comment by fatcerberus** [comment]
Oh, I didn’t realize the number and symbol cases behaved differently.  *All three cases* should definitely be an error, for the same reason.

@RyanCavanaugh ping

**Comment by potzkovge** [comment]
> @potzkovge The example given by @RyanCavanaugh makes sense in my eyes, as well. However, please have a look at my original question: why is there a difference between strings, numbers and symbols?
> 
> If the string case (createRecord1) is correct, the others (createRecord2 and createRecord3) must apparently be incorrect. Or not?

This inconsistency is exactly the possible bug. It may very well be that string is behaving correctly as stated a Record has to have all keys and if it is generic and unknown it should display the error. But then it should also display the exact same error for number and symbol. But i guess the confusion is around generics in this case.
